### PR TITLE
a8n: Implement `cancelCampaignPlan` mutation

### PR DESCRIPTION
**Pull request by mrnugget**
This implements the `cancelCampaignPlan` mutation that marks a `CampaignPlan` as cancelled in its `BackgroundProcessStatus` and stops all enqueued `CampaignJob`s from running.

It does this by
- adding a `canceled_at` field on `campaign_plans` and setting that in `cancelCampaignPlan` in case it hasn't been set.
- checking whether the the `CampaignPlan` has been canceled before executing a single `CampaignJob`. If it has, it marks the `CampaignJob` as errored.
- extending the `BackgroundProcessStatus` (in the GraphQL API and in the backend) with a new `CANCELED` state that's returned when a `CampaignPlan` has been executed.

 **Note**: this is not the most performant solution since we still do a `SELECT` and `UPDATE` for every `CampaignJob` if the `CampaignPlan` has been cancelled. But it's by far the most straightforward and easiest to debug solution.

Alternatives:
1. stop the workers in `Runner` once the plan has been executed. But in that case we'd need to do some concurrency foo to mark the `WaitGroup` as done.
2. add a `canceled_at` on `campaign_jobs`, set it along with `campaign_plan.canceled_at` and then do a `SELECT` (no `UPDATE`) to check whether it's been set before executing a job.

I think it's fine as it is and the `SELECT/UPDATEs` won't be a problem for a long time, but happy about suggestions :)

### CODE REVIEW

File: enterprise/internal/a8n/store.go [hunk 1]
```diff
@@ -1366,20 +1373,32 @@ func getCampaignPlanQuery(opts *GetCampaignPlanOpts) *sqlf.Query {
...
+WITH jobs AS (
+  SELECT
+    campaign_plan_id,
+    COUNT(*) AS total,
+    COUNT(*) FILTER (WHERE finished_at IS NULL) AS pending,
+    COUNT(*) FILTER (WHERE finished_at IS NOT NULL AND (diff != '' OR error != '')) AS completed,
+    array_agg(error) FILTER (WHERE error != '') AS errors
+  FROM campaign_jobs
+  GROUP BY campaign_plan_id
+)
```
Comment (felixfbecker) [inline enterprise/internal/a8n/store.go]: Is there a reason this has to be in a CTE instead of a sub query? 🤔

Comment (mrnugget) [reply enterprise/internal/a8n/store.go]: I liked the CTE more than the subquery.

File: cmd/frontend/graphqlbackend/schema.graphql [hunk 2]
```diff
@@ -495,6 +495,8 @@ enum BackgroundProcessState {
     ERRORED
     # The background process completed processing all items successfully.
     COMPLETED
+    # The background process was canceled.
+    CANCELED
```
Comment (ryanslade) [inline cmd/frontend/graphqlbackend/schema.graphql]: This looked wrong to me and now I know why:
https://www.grammarly.com/blog/canceled-vs-cancelled/

File: enterprise/internal/a8n/resolvers/resolver.go [hunk 3]
```diff
@@ -480,11 +481,25 @@ func (r *Resolver) CancelCampaignPlan(ctx context.Context, args graphqlbackend.C
...
-	return &graphqlbackend.EmptyResponse{}, nil
+	defer tx.Done(&err)
+
+	plan, err := tx.GetCampaignPlan(ctx, ee.GetCampaignPlanOpts{ID: id})
+	if err != nil {
+		return nil, err
+	}
+
+	if !plan.CanceledAt.IsZero() {
+		return nil, errors.New("execution has already been canceled")
```
Comment (ryanslade) [inline enterprise/internal/a8n/resolvers/resolver.go]: I wonder if this should be an error? If it's already been cancelled should we not just return nil?

Comment (mrnugget) [reply enterprise/internal/a8n/resolvers/resolver.go]: Yeah, I was unsure about that. But empty response seems better.

Should I do the same thing when a plan has been completed?

File: cmd/frontend/graphqlbackend/schema.graphql (hunk 2 above)
Comment (mrnugget) [reply cmd/frontend/graphqlbackend/schema.graphql]: haha! It looks incredibly wrong to me, I've only ever used `cancelled_at` in the past. But in the Go world `canceled` seems to be prevalent. See the documentation for `context` for example: https://golang.org/pkg/context/

File: enterprise/internal/a8n/resolvers/resolver.go (hunk 3 above)
Comment (ryanslade) [reply enterprise/internal/a8n/resolvers/resolver.go]: I think so. If the plan is in a state where cancelling wouldn't do anything then it should be a noop

Comment (mrnugget) [reply enterprise/internal/a8n/resolvers/resolver.go]: I added a commit that implements this.

File: enterprise/internal/a8n/store.go (hunk 1 above)
Comment (felixfbecker) [reply enterprise/internal/a8n/store.go]: I read that CTEs until the most recent Postgres version (which I think we don't use yet) get materialized and act as an optimization boundary - is that not a concern here?

Comment (mrnugget) [reply enterprise/internal/a8n/store.go]: We're talking about _hundreds_ of campaign jobs here. No, database performance was not a concern of mine here. If you have that concern, we could've saved 2 days of back-and-forth if that was in the first message.

**Pull request merged**